Lib input terminal

nuggreat opened the pull request:
This is a library for facilitating terminal input using direct key input as apposed to the on-screen keyboard form of `lib_input_string.ks` uses.

The documentation is not done yet as I would like some feedback on the feature set I already have before I start work on the documentation for the lib it.

scimas commented:
This is good. I had in mind an equivalent of "read_line" function found in the standard library of many languages. It would only get input from the user, accounting for the user moving cursor and deleting characters, print the characters as they are typed and return the string on enter key. It would also not limit the maximum number of characters in the string (that would make sense if kOS was more C - like rather than Python - like) or at least it would be a large number by default.
The number only input could be separate function.

nuggreat commented:
A single line input is more or less what you get with the string input.  The maxlength is to make it easy for people to do nicer terminal layouts and not have to worry about overrun.  The reason why I added a number only mode is because of some of my experience with with kOS GUIs and how bad the `:TOSCALAR()` suffix is at converting to a number if there is any thing at all that throws off the parse it gives up on the conversion and ether crashes or goes with the default answer.  It shouldn't be to hard to split them into different functions the only reason they the same is the commonalities between the two types of input for the actual concatenation steps.

scimas commented:
> The maxlength is to make it easy for people to do nicer terminal layouts

That is a good point, but it would be nice to make it an optional parameter with large default value for the string only version. It could be a mandatory parameter for the number only version where the max length also makes sense in regards to how big of a number the program is expecting.

nuggreat commented:
I have refactored the main lib to spit the input into several functions, added a second simpler example, and got the first draft of the documentation done.

scimas commented on doc/lib_input_terminal.md:
```diff
@@ -0,0 +1,81 @@
+  A string
+
+description:
+  This function takes a limited set of key presses on the kOS terminal and adds them into a single string.  As this function is intended for the input of scalars it only adds to a limited set of key presses, the keys it will add to the string are "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", and "." with "+" and "-" as additional control characters.
```
```suggestion
  This function takes key presses on the kOS terminal and adds them into a single string for displaying a number.  As such this function only works with a limited set of keys. The keys it will add to the string are "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", and "." with "+" and "-" for denoting sign.
```

scimas commented on doc/lib_input_terminal.md:
```diff
@@ -0,0 +1,81 @@
+description:
+  This function takes a limited set of key presses on the kOS terminal and adds them into a single string.  As this function is intended for the input of scalars it only adds to a limited set of key presses, the keys it will add to the string are "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", and "." with "+" and "-" as additional control characters.
+  While the string is being added to it will be printed on the terminal at the location set by the arguments.
+  On return there is the function will try to format the string so that is should be directly usable by the `:TOSCALAR()` suffix should the string fail it defaults to returning " 0"
```
```suggestion
  The function will also try to format the string so that it can be directly converted into a scalar by the `:TOSCALAR()` suffix. Should the formatting fail, " 0" is returned by default.
```

scimas commented on doc/lib_input_terminal.md:
```diff
@@ -0,0 +1,81 @@
+  This function takes a limited set of key presses on the kOS terminal and adds them into a single string.  As this function is intended for the input of scalars it only adds to a limited set of key presses, the keys it will add to the string are "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", and "." with "+" and "-" as additional control characters.
+  While the string is being added to it will be printed on the terminal at the location set by the arguments.
+  On return there is the function will try to format the string so that is should be directly usable by the `:TOSCALAR()` suffix should the string fail it defaults to returning " 0"
+  There is only allowed to be one "." character in a string at a time if a second "." is entered it will be ignored.
```
```suggestion
  There is only allowed to be one "." character in a string at a time. If a second "." is entered it will be ignored.
```

scimas commented on doc/lib_input_terminal.md:
```diff
@@ -0,0 +1,81 @@
+description:
+  This function reads the key presses and stores the current string until the loop is ended by pressing enter.
+  While the current string is being added to it will be printed on the terminal at the location set by the arguments.
+  The concatenator function gets passed the current string, non-control key presses, and maxLength.
```
```suggestion
  The concatenator function is called with the current string, non-control key presses, and maxLength.
```

scimas commented on doc/lib_input_terminal.md:
```diff
@@ -0,0 +1,81 @@
+  This function reads the key presses and stores the current string until the loop is ended by pressing enter.
+  While the current string is being added to it will be printed on the terminal at the location set by the arguments.
+  The concatenator function gets passed the current string, non-control key presses, and maxLength.
+    It is responsible for the rules of what a given key press will do do the current string and keeping the string from getting larger than maxLength.  It is expected to return the result of applying a key press back to `input_loop` to be stored as the current string.
```
```suggestion
    It is responsible for the rules of what a given key press will do to the current string and keeping the string from getting larger than maxLength.  It is expected to return the processed string back to `input_loop` - to be stored as the current string.
```

scimas commented:
I tested it today, looks good.

nuggreat commented:
That should be everything, I expanded `number_protect` as I remembered a bug in `:TOSCALAR()` because with large integer values see this kOS [issue](https://github.com/KSP-KOS/KOS/issues/2304) for details

scimas commented on SUMMARY.md:
```diff
@@ -33,11 +33,12 @@ This is an attempt at summarising and categorising the libraries by their use ca
-8. [lib_seven_seg.ks](https://github.com/KSP-KOS/KSLib/blob/master/library/lib_seven_seg.ks) / [docs](https://github.com/KSP-KOS/KSLib/blob/master/doc/lib_seven_seg.md) : creates a seven segment display to show single digit numbers.
-9. [lib_str_to_num.ks](https://github.com/KSP-KOS/KSLib/blob/master/library/lib_str_to_num.ks) / [docs](https://github.com/KSP-KOS/KSLib/blob/master/doc/lib_str_to_num.md) : converts numerical strings into numbers. This has been superseded by kOS built-in `STRING:TONUMBER()`.
-10. [lib_term.ks](https://github.com/KSP-KOS/KSLib/blob/master/library/lib_term.ks) / [docs](https://github.com/KSP-KOS/KSLib/blob/master/doc/lib_term.md) : a primitive terminal ASCII graphics library. Helps with drawing straight lines, circles and elliptical arcs.
+3. [lib_input_terminal](https://github.com/KSP-KOS/KSLib/blob/master/library/lib_input_terminal.ks) / [docks](https://github.com/KSP-KOS/KSLib/blob/master/doc/lib_input_terminal.md) : for string input by typing in the terminal.
```
```suggestion
3. [lib_input_terminal](https://github.com/KSP-KOS/KSLib/blob/master/library/lib_input_terminal.ks) / [docs](https://github.com/KSP-KOS/KSLib/blob/master/doc/lib_input_terminal.md) : for string input by typing in the terminal.

```

scimas commented on doc/lib_input_terminal.md:
```diff
@@ -0,0 +1,81 @@
+  A string
+  
+description:
+  This function takes key presses on the kOS terminal and adds them into a single string until the user presses enter to end return the string.
```
```suggestion
  This function takes key presses on the kOS terminal and adds them into a single string until the user presses enter to stop and return the string.
```
